core/domain/platform_parameter_domain.py

```diff
@@ -35,3 +67,976 @@ class PlatformParameterChange(change_domain.BaseChange):
+
+        return matched
+
+    def validate(self):
+        """Validates the PlatformParameterFilter domain object."""
+        if self._type not in self.SUPPORTED_FILTER_TYPE:
+            raise utils.ValidationError(
+                'Unsupported filter type %s' % self._type)
+
+        for op, _ in self._conditions:
+            if op not in self.SUPPORTED_OP_FOR_FILTERS[self._type]:
+                raise utils.ValidationError(
+                    'Unsupported comparison operator \'%s\' for %s filter, '
+                    'expected one of %s.' % (
+                        op, self._type,
+                        self.SUPPORTED_OP_FOR_FILTERS[self._type]))
+
+        if self._type == 'server_mode':
+            for _, mode in self._conditions:
+                if mode not in ALLOWED_SERVER_MODES:
+                    raise utils.ValidationError(
+                        'Invalid server mode %s, must be one of %s.' % (
+                            mode, ALLOWED_SERVER_MODES))
+        elif self._type == 'user_locale':
+            for _, locale in self._conditions:
+                if locale not in ALLOWED_USER_LOCALES:
+                    raise utils.ValidationError(
+                        'Invalid user locale %s, must be one of %s.' % (
+                            locale, ALLOWED_USER_LOCALES))
+        elif self._type == 'client_type':
+            for _, client_type in self._conditions:
+                if client_type not in ALLOWED_CLIENT_TYPES:
+                    raise utils.ValidationError(
+                        'Invalid client type %s, must be one of %s.' % (
+                            client_type, ALLOWED_CLIENT_TYPES))
+        elif self._type == 'app_version':
+            for _, version in self._conditions:
+                if not APP_VERSION_WITHOUT_HASH_REGEXP.match(version):
+                    raise utils.ValidationError(
+                        'Invalid version expression %s, expected to match'
+                        'regexp %s.' % (
+                            version, APP_VERSION_WITHOUT_HASH_REGEXP))
+
+    def to_dict(self):
+        """Returns a dict representation of the PlatformParameterFilter domain
+        object.
+
+        Returns:
+            dict. A dict mapping of all fields of PlatformParameterFilter
+            object.
+        """
+        return {
+            'type': self._type,
+            'conditions': self._conditions,
+        }
+
+    @classmethod
+    def from_dict(cls, filter_dict):
+        """Returns an PlatformParameterFilter object from a dict.
+
+        Args:
+            filter_dict: dict. A dict mapping of all fields of
+                PlatformParameterFilter object.
+
+        Returns:
+            PlatformParameterFilter. The corresponding PlatformParameterFilter
+            domain object.
+        """
+        return cls(filter_dict['type'], filter_dict['conditions'])
+
+    @classmethod
+    def _match_version_expression(cls, op, value, client_version):
```

Comment:
Why is this a class method? You even call it from the self context.

Reply:
The reason was that this method does not use any instance information (similar reason for `_is_first_version_smaller`).
Changed `_match_version_expression` `_is_first_version_smaller` to instance method for clarity.